node/src/certificate_store.rs

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crypto::Hash;
+use std::collections::{HashMap, VecDeque};
+use std::sync::Arc;
+use store::rocks::DBMap;
+use store::rocks::TypedStoreError::RocksDBError;
+use store::Map;
+use tokio::sync::{oneshot, RwLock};
+use types::{Certificate, CertificateDigest, Round, StoreResult};
+
+/// The main storage when we have to deal with certificates. It maintains
+/// two storages, one main which saves the certificates by their ids, and a
+/// secondary one which acts as an index to allow us fast retrieval based
+/// for queries based in certificate rounds.
+#[derive(Clone)]
+pub struct CertificateStore {
+    /// Holds the certificates by their digest id
+    certificates_by_id: DBMap<CertificateDigest, Certificate>,
+    /// A secondary index that keeps the certificate digest ids
+    /// by the certificate rounds. That helps us to perform
+    /// range requests based on rounds. We avoid storing again the
+    /// certificate here to not waste space. To dereference we use
+    /// the certificates_by_id storage.
+    certificate_ids_by_round: DBMap<(Round, CertificateDigest), u8>,
+    /// Senders to notify for a write that happened for
+    /// the specified certificate digest id
+    obligations: Arc<RwLock<HashMap<CertificateDigest, VecDeque<oneshot::Sender<Certificate>>>>>,
```

Comment:
I could instead use a [DashMap](https://crates.io/crates/dashmap) which you have also used in the dag implementation

Reply:
My intention with having interior mutability here is to be able to share/clone the `CertificateStore` across structs - the "penalty" is having a lock around the obligations I guess, but it might not be that much of an issue